src/v/ssx/rwlock.h

```diff
@@ -0,0 +1,169 @@
+    static constexpr Clock::duration default_log_timeout = 10000ms;
+
+    struct holder {
+    public:
+        holder() = delete;
+        holder(const holder&) = delete;
+
+        holder(logging_rwlock* lock, ss::rwlock::holder&& h, holder_id id)
+          : _lock(lock)
+          , _h(std::move(h))
+          , _id(id) {}
+        holder(holder&& o) noexcept
+          : _lock(std::exchange(o._lock, nullptr))
+          , _h(std::move(o._h))
+          , _id(o._id) {}
+        holder& operator=(holder&& o) noexcept {
+            _lock = std::exchange(o._lock, nullptr);
+            _h = std::move(o._h);
+            _id = o._id;
+            return *this;
+        }
+        ~holder() {
+            if (_lock) {
+                vlog(
+                  _lock->_logger.trace,
+                  "Dropping lock on {} ({})",
+                  _lock->_resource_str_gen(),
+                  _id);
+            }
+        }
+
+    private:
+        logging_rwlock* _lock;
+        ss::rwlock::holder _h;
+        holder_id _id;
+    };
+
+    explicit logging_rwlock(ss::logger& logger)
+      : _logger(logger) {}
+
+    explicit logging_rwlock(ss::logger& logger, resource_str_func fn)
+      : _logger(logger)
+      , _resource_str_gen(std::move(fn)) {}
+
+    ss::future<holder> hold_read_lock(
+      ss::semaphore::time_point take_deadline
+      = ss::semaphore::time_point::max(),
+      Clock::duration log_timeout = default_log_timeout) {
+        return lock_with_log(
+          [this, take_deadline]() {
+              return _lock.hold_read_lock(take_deadline);
+          },
+          "Read",
+          log_timeout);
+    }
+
+    ss::future<holder> hold_write_lock(
+      ss::semaphore::time_point take_deadline
+      = ss::semaphore::time_point::max(),
+      Clock::duration log_timeout = default_log_timeout) {
+        return lock_with_log(
+          [this, take_deadline]() {
+              return _lock.hold_write_lock(take_deadline);
+          },
+          "Write",
+          log_timeout);
+    }
+
+    bool locked() const { return _lock.locked(); }
+
+    bool try_write_lock() {
+        auto ret = _lock.try_write_lock();
+        if (ret) {
+            vlog(
+              _logger.trace,
+              "Write lock on {} acquired immediately",
+              _resource_str_gen());
+        }
+        return ret;
+    }
+
+    void write_unlock() {
+        vlog(_logger.trace, "Dropping write lock on {}", _resource_str_gen());
+        _lock.write_unlock();
+    }
+
+private:
+    friend struct holder;
+
+    template<typename F>
+    requires ReturnsFutureHolder<F> ss::future<holder> lock_with_log(
+      F lock_func, const char* acquire_type, Clock::duration log_timeout) {
+        auto id = _next_id++;
+        vlog(
+          _logger.trace,
+          "{} lock on {} ({}) being acquired",
+          acquire_type,
+          _resource_str_gen(),
+          id);
+        ss::timer<Clock> log_timer;
```

Comment:
trying to understand the lifetime of log_timer... doesn't the dtor of `holder` need to own the log_timer ? 

why not have a 


```
struct holder {
   holder() {
       initialize the timer here
    
   }
   ~holder(){
       reset/disarm the timer here 
   }

};


```
also i wonder if you can do 


```
struct holder: final ss::timer {
  holder() {
     init the timer here
  }

};

```

anyway, was driving by and got nerd sniped, so feel free to disregard.

Reply:
ahhhhh... answer to my question is you do 

```coawait lock_func()````

Reply:
Yeah, also worth noting that the timer here isn't really for timing how long we hold -- it's for timing how long it takes to acquire the lock, so that we can error log when it's something unreasonable. I'd considered also logging timing how long we hold, but with logs on hand that's something we can calculate based on log timestamps